// avalanche/typings/src/apis/avm/credentials.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NFTCredential = exports.SecpCredential = exports.Credential = exports.SelectCredentialClass = void 0;
/**
 * @packageDocumentation
 * @module AVMAPI-Credentials
 */
const buffer_1 = require("buffer/");
const bintools_1 = __importDefault(require("../../utils/bintools"));
const types_1 = require("./types");
/**
 * @ignore
 */
const bintools = bintools_1.default.getInstance();
/**
 * Takes a buffer representing the credential and returns the proper [[Credential]] instance.
 *
 * @param credid A number representing the credential ID parsed prior to the bytes passed in
 *
 * @returns An instance of an [[Credential]]-extended class.
 */
exports.SelectCredentialClass = (credid, ...args) => {
    if (credid === types_1.AVMConstants.SECPCREDENTIAL) {
        const secpcred = new SecpCredential(...args);
        return secpcred;
    }
    if (credid === types_1.AVMConstants.NFTCREDENTIAL) {
        const nftcred = new NFTCredential(...args);
        return nftcred;
    }
    /* istanbul ignore next */
    throw new Error(`Error - SelectCredentialClass: unknown credid ${credid}`);
};
class Credential {
    constructor(sigarray = undefined) {
        this.sigArray = [];
        /**
           * Adds a signature to the credentials and returns the index off the added signature.
           */
        this.addSignature = (sig) => {
            this.sigArray.push(sig);
            return this.sigArray.length - 1;
        };
        if (typeof sigarray !== 'undefined') {
            /* istanbul ignore next */
            this.sigArray = sigarray;
        }
    }
    fromBuffer(bytes, offset = 0) {
        const siglen = bintools.copyFrom(bytes, offset, offset + 4).readUInt32BE(0);
        offset += 4;
        this.sigArray = [];
        for (let i = 0; i < siglen; i++) {
            const sig = new types_1.Signature();
            offset = sig.fromBuffer(bytes, offset);
            this.sigArray.push(sig);
        }
        return offset;
    }
    toBuffer() {
        const siglen = buffer_1.Buffer.alloc(4);
        siglen.writeInt32BE(this.sigArray.length, 0);
        const barr = [siglen];
        let bsize = siglen.length;
        for (let i = 0; i < this.sigArray.length; i++) {
            const sigbuff = this.sigArray[i].toBuffer();
            bsize += sigbuff.length;
            barr.push(sigbuff);
        }
        return buffer_1.Buffer.concat(barr, bsize);
    }
}
exports.Credential = Credential;
class SecpCredential extends Credential {
    getCredentialID() {
        return types_1.AVMConstants.SECPCREDENTIAL;
    }
}
exports.SecpCredential = SecpCredential;
class NFTCredential extends Credential {
    getCredentialID() {
        return types_1.AVMConstants.NFTCREDENTIAL;
    }
}
exports.NFTCredential = NFTCredential;
