// avalanche/typings/tests/apis/admin/api.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_mock_axios_1 = __importDefault(require("jest-mock-axios"));
const src_1 = require("src");
describe('Admin', () => {
    const ip = '127.0.0.1';
    const port = 9650;
    const protocol = 'https';
    const username = 'AvaLabs';
    const password = 'password';
    const avalanche = new src_1.Avalanche(ip, port, protocol, 12345, 'What is my purpose? You pass butter. Oh my god.', undefined, false);
    let admin;
    beforeAll(() => {
        admin = avalanche.Admin();
    });
    afterEach(() => {
        jest_mock_axios_1.default.reset();
    });
    test('getNodeID', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.getNodeID();
        const payload = {
            result: {
                nodeID: 'abcd',
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe('abcd');
    }));
    test('getNodeVersion', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.getNodeVersion();
        const payload = {
            result: {
                version: 'avalanche/0.5.5',
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe('avalanche/0.5.5');
    }));
    test('alias', () => __awaiter(void 0, void 0, void 0, function* () {
        const ep = '/ext/something';
        const al = '/ext/anotherthing';
        const result = admin.alias(ep, al);
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test('aliasChain', () => __awaiter(void 0, void 0, void 0, function* () {
        const ch = 'abcd';
        const al = 'myChain';
        const result = admin.aliasChain(ch, al);
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test('getNetworkID', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.getNetworkID();
        const payload = {
            result: {
                networkID: 12345,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(12345);
    }));
    test('getBlockchainID', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.getBlockchainID('avm');
        const payload = {
            result: {
                blockchainID: avalanche.XChain().getBlockchainID(),
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe('What is my purpose? You pass butter. Oh my god.');
    }));
    test('getNetworkName', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.getNetworkName();
        const payload = {
            result: {
                networkName: 'denali',
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe('denali');
    }));
    test('lockProfile', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.lockProfile('filename');
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test('memoryProfile', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.memoryProfile('filename');
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test('peers', () => __awaiter(void 0, void 0, void 0, function* () {
        const peers = ['p1', 'p2'];
        const result = admin.peers();
        const payload = {
            result: {
                peers,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(peers);
    }));
    test('startCPUProfiler', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.startCPUProfiler('filename');
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
    test('stopCPUProfiler', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = admin.stopCPUProfiler();
        const payload = {
            result: {
                success: true,
            },
        };
        const responseObj = {
            data: payload,
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
});
